rpc_jobs/release.yml

```diff
@@ -0,0 +1,68 @@
+- job:
+    name: RE-Release
+    project-type: workflow
+    concurrent: false
+    properties:
+      - build-discarder:
+          num-to-keep: 30
+    parameters:
+      - rpc_gating_params
+      - string:
+          name: RELEASE_VERSION
+          description: |
+            The version that will be tagged by this job
+      - string:
+          name: NEXT_VERSION
+          description: |
+            The version that will be released after the next stabalisation
+            period.
+      - string:
+          name: MAINLINE
+          description: The mainline branch to work from
+      - string:
+          name: REPO_URL
+          description: |
+            URL of the working repo. Used to pull and push, so should be
+            git@github.com:owner/repo. Key for
+            https://github.com/rpc-jenkins-svc is used for pushing.
+      - string:
+          name: STAGES
+          description: |
+            Items to run. Options:
+              Tag,
+              Update RC Branch
+          default: >-
+            Tag,
+            Update RC Branch
+    dsl: |
+      stage("Check Parameters"){
+        if (env.REPO_URL == "" || env.MAINLINE == ""){
+          throw new Exception(
+            "REPO_URL and MAINLINE parameters are required.")
+        }
+        if (!(REPO_URL ==~ /git@github.com:[\w-_]*\/[\w-_]*/)){
+          throw new Exception(
+            "Repo url should be of the form git@github.com:owner/repo")
+        }
+        print ("Parameter check successful.")
+      }
+      library "rpc-gating@${RPC_GATING_BRANCH}"
+      common.shared_slave(){
+        rc = "${MAINLINE}-rc"
+        stage("Configure Git"){
+          release.configure_git()
+        }
+        stage("Clone"){
+          release.clone(REPO_URL)
+        }
+        common.conditionalStage(
+          stage_name: "Tag",
+          stage: {
+            release.tag(RELEASE_VERSION, rc)
+          })
+        common.conditionalStage(
+          stage_name: "Update RC Branch",
+          stage: {
+            release.update_rc_branch(REPO_URL, MAINLINE, rc)
```

Comment:
That's definitely clearer.

I don't think the order is correct and the naming conventions are confusing.
We should probably also separate these into different jobs, because some may trigger at different occasions (ad-hoc things).

Let's take a step back. What can happen (examples afterwards for rpc-o):
a) Tag <whatever> in a stabilisation branch with no artefact update (example, doc fix, or minor bug that doesn't require new artefacts, because not everything is an artefact yet).
In this case, the process should probably triggered by the team, and we shouldn't see, as RE, the difference. IMO if the team is managing their repo, we shouldn't care, and we shouldn't build a job for it.
b) Tag <whatever> in a stabilisation branch with artefact update (example, osa change, new version of ELK...). In that case, the team managing their repo would like to have a way to trigger an artifact update that's external to their power. That's where a job comes in, giving a reference/tag. It still doesn't matter for this PR, as this is only caring about RE branching strategies in a global way. Each of the teams have to manage their own work of what's happening in a new tag (and in the meantime, we can help them RPC-O with the setup/automation of this work).
c) Regular RPC-O release triggering, so no ad-hoc work.
That's where this job should trigger, and it should happen simply as this:

1. List all the mainlines we know for all the products (RPC-O, K8s, MaaS)
2. Train is at the station for YYYY-MM (giving an example for a monthly cycle, where MM is next month). Check if our release repo has a review from each of the mainlines into a YYYY-MM file. These PR probably list a tag + sha for their mainline. If a team has no patch, that team missed the train.
3. Wait for a time to determine, which is max(artefact build time).
4. Because all these commits listed in 2. are very similar to the case b) above, I think we can assume that these commits would have build artefacts and updated their products accordingly (that's why we are doing step 3).
5. Then the YYYY-MM document is our manifest of the artefacts that were build for YYYY-MM. When release time (=MM) is reached, release (publishing _publicly_ all the artefacts somewhere)

So I am confused. When I read "Release job", I thought we were thinking of step 5.
Here it seems more like it's kinda b) ?

But let's focus on 5) for this PR, the "manual release job". What should this do?
IMO, building next files for manifest, checking availability of artefacts, publishing them, NOT TAGGING(done in b or 2), cut a new "stabilisation" branch YYYY-MM+1 from tag listed in YYYY-MM or from head of their branch YYYY-MM.
This makes our life simpler: We just read/write files and cut branches.

Am I completely wrong? Separating the tagging of a version from the "Release train" seem easier to me.